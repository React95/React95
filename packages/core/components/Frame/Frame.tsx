import React, { forwardRef } from 'react';
import type {
  ComponentPropsWithRef,
  ElementType,
  ForwardedRef,
  Ref,
  ReactNode,
  RefAttributes,
} from 'react';
import { Sprinkles, sprinkles } from './Frame.css';
import cn from 'classnames';

type FixedForwardRef = <T, P = object>(
  render: (props: P, ref: Ref<T>) => ReactNode,
) => (props: P & RefAttributes<T>) => ReactNode;

export const fixedForwardRef = forwardRef as FixedForwardRef;

type DistributiveOmit<T, TOmitted extends PropertyKey> = T extends any
  ? Omit<T, TOmitted>
  : never;

export type FrameProps<TAs extends ElementType> = {
  as?: TAs;
} & DistributiveOmit<
  ComponentPropsWithRef<ElementType extends TAs ? 'div' : TAs>,
  'as'
> &
  Sprinkles;

const FrameComponent = <TAs extends ElementType = 'div'>(
  props: FrameProps<TAs>,
  ref: ForwardedRef<any>,
) => {
  const { as, children, ...rest } = props;
  const Component = as || 'div';
  const { className, style, otherProps } = sprinkles(rest);

  return (
    <Component
      {...otherProps}
      style={{ ...style, ...otherProps.style }}
      className={cn(className, otherProps.className)}
      ref={ref}
    >
      {children}
    </Component>
  );
};

export const Frame = fixedForwardRef(FrameComponent);
