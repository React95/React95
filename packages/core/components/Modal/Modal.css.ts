import { globalStyle, style } from '@vanilla-extract/css';
import { contract } from '../themes/contract.css';
import { recipe } from '@vanilla-extract/recipes';

export const modalWrapper = recipe({
  base: {
    display: 'flex',
    flexDirection: 'column',
    position: 'fixed',
    padding: contract.space[2],
    top: '50px',
    backgroundColor: contract.colors.material,
    boxShadow: contract.shadows.out,
  },
  variants: {
    active: {
      true: {
        zIndex: contract.zIndices.modal,
      },
    },
    minimized: {
      true: {
        display: 'none',
      },
      false: {
        display: 'flex',
      },
    },
  },
});

export const buttonWrapper = style({
  display: 'flex',
  gap: contract.space[6],
  flexDirection: 'row',
  padding: contract.space[6],
});

globalStyle(`${buttonWrapper} button`, {
  minWidth: '70px',
});

export const content = style({
  flexGrow: 1,
  display: 'flex',
  flexDirection: 'column',
  padding: contract.space[6],
  marginBlockStart: contract.space[4],
  marginInlineEnd: contract.space[1],
});

export const menuWrapper = style({
  display: 'flex',
  flexDirection: 'row',
  listStyle: 'none',
  margin: '0',
  paddingLeft: '0',
  paddingTop: '1px',
  paddingBottom: '1px',
  borderBottomStyle: 'solid',
  borderWidth: '1px',
  borderBottomColor: contract.colors.borderDark,
  boxShadow: `0 1px 0 0 ${contract.colors.borderLighter}`,
});

export const menuItem = recipe({
  base: {
    position: 'relative',
    paddingLeft: '6px',
    paddingRight: '6px',
    userSelect: 'none',
  },
  variants: {
    active: {
      true: {
        backgroundColor: contract.colors.material,
        color: contract.colors.materialTextInvert,
      },
    },
  },
});

globalStyle(`${menuItem.classNames.base} ul`, {
  position: 'absolute',
  left: '0',
  color: contract.colors.materialText,
});
