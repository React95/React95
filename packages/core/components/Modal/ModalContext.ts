/* eslint-disable no-unused-vars */
import { createContext } from 'react';

export type Windows = {
  icon?: React.ReactElement;
  title: string;
  hasButton: boolean;
};

export interface IModalContextProps {
  windows: Record<string, Windows>;
  addWindows(window: Windows): string | void;
  removeWindow(id: string): void;
  updateWindow(id: string, window: Windows): void;
  setActiveWindow(title: string): void;
  activeWindow?: string;
}

export const ModalContext = createContext<IModalContextProps>({
  windows: {},
  addWindows: () => {},
  removeWindow: () => {},
  updateWindow: () => {},
  setActiveWindow: () => {},
  activeWindow: '',
});
