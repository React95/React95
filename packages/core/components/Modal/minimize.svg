<svg xmlns="http://www.w3.org/2000/svg" width="9" height="9"><g class="layer"><path d="M0 6.95h9v2.04H0V6.95z"/></g></svg>