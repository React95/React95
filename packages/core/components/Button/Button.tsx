import React from 'react';
import type {
  ElementType,
  ReactElement,
  ElementRef,
  ForwardedRef,
  ButtonHTMLAttributes,
} from 'react';
import cn from 'classnames';

import { button } from './Button.css';
import { Frame, FrameProps, fixedForwardRef } from '../Frame/Frame';

export type ButtonProps<TAs extends ElementType> =
  ButtonHTMLAttributes<HTMLButtonElement> & FrameProps<TAs>;

export const Button = fixedForwardRef<HTMLButtonElement, ButtonProps<'button'>>(
  (props, ref) => {
    return (
      <Frame
        as="button"
        {...props}
        className={cn(button, props.className)}
        ref={ref}
      />
    );
  },
) as <TAs extends ElementType = 'button'>(
  props: ButtonProps<TAs> & { ref?: ForwardedRef<ElementRef<TAs>> },
) => ReactElement;
