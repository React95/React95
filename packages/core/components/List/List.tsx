import React, { forwardRef } from 'react';
import cn from 'classnames';

import { ListItem } from './ListItem';
import { Divider } from './ListDivider';
import { Frame, FrameProps } from '../Frame/Frame';
import { list } from './List.css';

export type ListProps = Omit<FrameProps<'ul'>, 'as'>;

const ListRenderer = forwardRef<HTMLUListElement, ListProps>((rest, ref) => (
  <Frame {...rest} ref={ref} className={cn(list, rest.className)} as="ul" />
));

export type IListProps = typeof ListRenderer & {
  Item: typeof ListItem;
  Divider: typeof Divider;
};

export const List: IListProps = Object.assign(ListRenderer, {
  Item: ListItem,
  Divider: Divider,
});
