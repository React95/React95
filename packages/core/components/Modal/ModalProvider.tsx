import { nanoid } from 'nanoid';
import * as React from 'react';

import { ModalContext, Windows } from './ModalContext';

// eslint-disable-next-line @typescript-eslint/ban-types
export type ModalProviderProps = {};

type WindowStack = Record<string, Windows>;

type WindowAction =
  | {
      type: 'ADD_WINDOW';
      id: string;
      window: Windows;
    }
  | {
      type: 'UPDATE_WINDOW';
      id: string;
      window: Windows;
    }
  | {
      type: 'REMOVE_WINDOW';
      id: string;
    };

const windowStackReducer: React.Reducer<WindowStack, WindowAction> = (
  state,
  action,
) => {
  const newWindows = { ...state };
  switch (action.type) {
    case 'ADD_WINDOW': {
      newWindows[action.id] = action.window;
      return newWindows;
    }
    case 'REMOVE_WINDOW': {
      delete newWindows[action.id];
      return newWindows;
    }
    case 'UPDATE_WINDOW': {
      newWindows[action.id] = action.window;
      return newWindows;
    }
    default:
      return state;
  }
};

export const ModalProvider: React.FC<
  React.PropsWithChildren<ModalProviderProps>
> = ({ children }) => {
  const [windows, dispatch] = React.useReducer(windowStackReducer, {});
  const [activeWindow, setActiveWindow] = React.useState<string>();

  const addWindows = (window: Windows) => {
    const id = nanoid();
    dispatch({ type: 'ADD_WINDOW', id, window });

    return id;
  };
  const removeWindow = (id: string) => {
    dispatch({ type: 'REMOVE_WINDOW', id });
  };
  const updateWindow = (id: string, window: Windows) => {
    dispatch({ type: 'UPDATE_WINDOW', id, window });
  };

  return (
    <ModalContext.Provider
      value={{
        windows,
        addWindows,
        removeWindow,
        updateWindow,
        setActiveWindow,
        activeWindow,
      }}
    >
      {children}
    </ModalContext.Provider>
  );
};
