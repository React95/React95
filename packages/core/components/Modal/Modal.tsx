import { nanoid } from 'nanoid';
import React, {
  HTMLAttributes,
  MouseEvent,
  ReactElement,
  Ref,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';

import { DragOptions, useDraggable } from '@neodrag/react';

import { Button } from '../Button/Button';
import { fixedForwardRef, Frame, FrameProps } from '../Frame/Frame';
import { List } from '../List/List';
import { TitleBar, TitleBarBackgroundProps, OptionProps } from '../TitleBar/TitleBar';
import * as styles from './Modal.css';

import cn from 'classnames';
import { useOnClickOutside } from 'usehooks-ts';
import { ModalEvents, modals } from '../shared/events';

export type ModalButtons = {
  value: string;
  onClick(event: MouseEvent): void;
};

export type ModalMenu = {
  name: string;
  list: ReactElement<typeof List>;
};

export type ModalDefaultPosition = {
  x: number;
  y: number;
};

type TitleBarOptions =
  | typeof TitleBar.Close
  | typeof TitleBar.Help
  | typeof TitleBar.Maximize
  | typeof TitleBar.Minimize
  | typeof TitleBar.Restore;

export type ModalProps = {
  buttons?: Array<ModalButtons>;
  menu?: Array<ModalMenu>;
  dragOptions?: Omit<DragOptions, 'handle'>;
  hasWindowButton?: boolean;
  buttonsAlignment?: FrameProps<'div'>['justifyContent'];
  titleBarOptions?:
    | ReactElement<TitleBarOptions>
    | ReactElement<TitleBarOptions>[];
} & Omit<FrameProps<'div'>, 'as'> &
  HTMLAttributes<HTMLDivElement> &
  Pick<TitleBarBackgroundProps, 'title' | 'icon'>;

const ModalContent = fixedForwardRef<HTMLDivElement, FrameProps<'div'>>(
  (rest, ref) => (
    <Frame {...rest} ref={ref} className={cn(styles.content, rest.className)} />
  ),
);

const ModalMinimize = fixedForwardRef<HTMLButtonElement, OptionProps<'button'>>((props, ref) => {
  const [id, setId] = useState<string>("");

  useEffect(() => {
    const handleVisibilityChange = ({ id: activeId }: { id: string }) => {
      setId(activeId); 
    };

    modals.on(ModalEvents.ModalVisibilityChanged, handleVisibilityChange);

    return () => {
      modals.off(ModalEvents.ModalVisibilityChanged, handleVisibilityChange);
    };
  }, []);

  const handleMinimize = () => {
    modals.emit(ModalEvents.MinimizeModal, { id });
    modals.emit(ModalEvents.ModalVisibilityChanged, { id: 'no id' });
  };

  return <TitleBar.Minimize {...props} ref={ref} onClick={handleMinimize} />;
});

const ModalRenderer = (
  {
    hasWindowButton: hasButton = true,
    buttons = [],
    buttonsAlignment = 'flex-end',
    children,
    icon,
    menu = [],
    title,
    dragOptions,
    titleBarOptions,
    className,
    ...rest
  }: ModalProps,
  ref: Ref<HTMLDivElement | null>,
) => {
  const [id] = useState<string>(nanoid());
  const [menuOpened, setMenuOpened] = useState('');
  const [isActive, setIsActive] = useState(false);
  const [isModalMinimized, setIsModalMinimized] = useState(false);

  const draggableRef = useRef<HTMLDivElement>(null);
  useDraggable(draggableRef, {
    ...dragOptions,
    handle: '.draggable',
  });

  const menuRef = useRef<HTMLUListElement>(null);
  useOnClickOutside(menuRef, () => {
    setMenuOpened('');
  });

  useEffect(() => {
    modals.emit(ModalEvents.AddModal, {
      icon,
      title,
      id,
      hasButton,
    });

    modals.on(ModalEvents.ModalVisibilityChanged, ({ id: activeId }) => {
      setIsActive(activeId === id);
    });

    modals.emit(ModalEvents.ModalVisibilityChanged, { id });

    return () => {
      modals.emit(ModalEvents.RemoveModal, { id });
    };
  }, []);

  useEffect(() => {
    modals.on(ModalEvents.MinimizeModal, ({ id: activeId }) => {
      if (activeId === id) {
        setIsModalMinimized(true);
      }
    });

    modals.on(ModalEvents.RestoreModal, ({ id: activeId }) => {
      if (activeId === id) {
        setIsModalMinimized(false);
      }
    });

    return () => {
      modals.off(ModalEvents.MinimizeModal, () => {});
      modals.off(ModalEvents.RestoreModal, () => {});
    };
  }, [id]);

  useImperativeHandle(ref, () => {
    return draggableRef.current;
  });

  return (
    <Frame
      {...rest}
      className={cn(
        styles.modalWrapper({ active: isActive, minimized: isModalMinimized }),
        className,
      )}
      ref={draggableRef}
      onMouseDown={() => {
        modals.emit(ModalEvents.ModalVisibilityChanged, { id });
      }}
    >
      <TitleBar
        active={isActive}
        icon={icon}
        title={title}
        className="draggable"
      >
        {titleBarOptions && (
          <TitleBar.OptionsBox>{titleBarOptions}</TitleBar.OptionsBox>
        )}
      </TitleBar>

      {menu && menu.length > 0 && (
        <ul className={styles.menuWrapper} ref={menuRef}>
          {menu.map(({ name, list }) => {
            const active = menuOpened === name;
            return (
              <li
                key={name}
                onMouseDown={() => setMenuOpened(name)}
                className={styles.menuItem({ active })}
              >
                {name}
                {active && list}
              </li>
            );
          })}
        </ul>
      )}

      {children}

      {buttons && buttons.length > 0 && (
        <Frame
          className={styles.buttonWrapper}
          justifyContent={buttonsAlignment}
        >
          {buttons.map(button => (
            <Button
              key={button.value}
              onClick={button.onClick}
              value={button.value}
            >
              {button.value}
            </Button>
          ))}
        </Frame>
      )}
    </Frame>
  );
};

export const Modal = Object.assign(
  fixedForwardRef<HTMLDivElement, ModalProps>(ModalRenderer),
  {
    Content: ModalContent,
    Minimize: ModalMinimize,
  },
) as ModalProps &
  typeof ModalRenderer & {
    Content: typeof ModalContent;
    Minimize: typeof ModalMinimize;
  };
