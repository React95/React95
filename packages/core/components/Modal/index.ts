import { Modal } from './Modal';
import ModalContext from './ModalContext';
import ModalProvider from './ModalProvider';

export { Modal, ModalContext, ModalProvider };
