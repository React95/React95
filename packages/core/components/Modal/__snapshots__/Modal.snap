// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Modal /> > Snapshots > should match snapshot 1`] = `
<div>
  <div
    class="Modal_modalWrapper__1txblt60 Modal_modalWrapper_active_true__1txblt61 Modal_modalWrapper_minimized_false__1txblt63 neodrag"
    style="transform: translate3d(0px, 0px, 0); touch-action: none;"
  >
    <div
      class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_true__1kpzw683 draggable"
    >
      <svg test-file-stub />
      <div
        class="TitleBar_title__1kpzw685"
      >
        file.bat
      </div>
    </div>
    <ul
      class="Modal_menuWrapper__1txblt66"
    >
      <li
        class="Modal_menuItem__1txblt67"
      >
        File
      </li>
      <li
        class="Modal_menuItem__1txblt67"
      >
        Edit
      </li>
    </ul>
    <div
      class="Modal_content__1txblt65"
    >
      Hello
    </div>
    <div
      class="Frame_justifyContent__iuzg891f Modal_buttonWrapper__1txblt64"
      style="--justifyContent__iuzg891e: flex-end;"
    >
      <button
        class="Button_button__1bzq74l0"
        value="Ok"
      >
        Ok
      </button>
      <button
        class="Button_button__1bzq74l0"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`<Modal /> > Snapshots > should match snapshot with buttonsAligment prop equals center 1`] = `
<div>
  <div
    class="Modal_modalWrapper__1txblt60 Modal_modalWrapper_active_true__1txblt61 Modal_modalWrapper_minimized_false__1txblt63 neodrag"
    style="transform: translate3d(0px, 0px, 0); touch-action: none;"
  >
    <div
      class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_true__1kpzw683 draggable"
    >
      <svg test-file-stub />
      <div
        class="TitleBar_title__1kpzw685"
      >
        file.bat
      </div>
    </div>
    <div
      class="Modal_content__1txblt65"
    >
      Hello
    </div>
    <div
      class="Frame_justifyContent__iuzg891f Modal_buttonWrapper__1txblt64"
      style="--justifyContent__iuzg891e: center;"
    >
      <button
        class="Button_button__1bzq74l0"
        value="Ok"
      >
        Ok
      </button>
      <button
        class="Button_button__1bzq74l0"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`<Modal /> > Snapshots > should match snapshot with buttonsAligment prop equals flex-start 1`] = `
<div>
  <div
    class="Modal_modalWrapper__1txblt60 Modal_modalWrapper_active_true__1txblt61 Modal_modalWrapper_minimized_false__1txblt63 neodrag"
    style="transform: translate3d(0px, 0px, 0); touch-action: none;"
  >
    <div
      class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_true__1kpzw683 draggable"
    >
      <svg test-file-stub />
      <div
        class="TitleBar_title__1kpzw685"
      >
        file.bat
      </div>
    </div>
    <div
      class="Modal_content__1txblt65"
    >
      Hello
    </div>
    <div
      class="Frame_justifyContent__iuzg891f Modal_buttonWrapper__1txblt64"
      style="--justifyContent__iuzg891e: flex-start;"
    >
      <button
        class="Button_button__1bzq74l0"
        value="Ok"
      >
        Ok
      </button>
      <button
        class="Button_button__1bzq74l0"
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`<Modal /> > Snapshots > should match snapshot with titleBarOptions 1`] = `
<div>
  <div
    class="Modal_modalWrapper__1txblt60 Modal_modalWrapper_active_true__1txblt61 Modal_modalWrapper_minimized_false__1txblt63 neodrag"
    style="transform: translate3d(0px, 0px, 0); touch-action: none;"
  >
    <div
      class="TitleBar_titleBarBackground__1kpzw682 TitleBar_titleBarBackground_active_true__1kpzw683 draggable"
    >
      <svg test-file-stub />
      <div
        class="TitleBar_title__1kpzw685"
      >
        file.bat
      </div>
      <div
        class="TitleBar_optionsBox__1kpzw680"
      >
        <button
          class="Button_button__1bzq74l0 TitleBar_option__1kpzw681"
        >
          <img
            alt="help"
            src="test-file-stub"
          />
        </button>
        <button
          class="Button_button__1bzq74l0 TitleBar_option__1kpzw681"
        >
          <img
            alt="maximize"
            src="test-file-stub"
          />
        </button>
        <button
          class="Button_button__1bzq74l0 TitleBar_option__1kpzw681"
        >
          <img
            alt="minimize"
            src="test-file-stub"
          />
        </button>
        <button
          class="Button_button__1bzq74l0 TitleBar_option__1kpzw681"
        >
          <img
            alt="restore"
            src="test-file-stub"
          />
        </button>
        <button
          class="Button_button__1bzq74l0 TitleBar_option__1kpzw681"
        >
          <img
            alt="close"
            src="test-file-stub"
          />
        </button>
      </div>
    </div>
    <div
      class="Modal_content__1txblt65"
    >
      Hello
    </div>
  </div>
</div>
`;
